subprojects/language-native/src/integTest/groovy/org/gradle/language/nativeplatform/ParallelNativePluginsIntegrationTest.groovy

```diff
@@ -140,5 +146,149 @@ class ParallelNativePluginsIntegrationTest extends AbstractInstalledToolChainInt
         installations.every { installation, app ->
             installation.exec().out == app.englishOutput
         }
+
+        and:
+        buildOperations.assertConcurrentOperationsExecuted(ExecuteTaskBuildOperationType)
+    }
+
+    def "can execute link executable tasks in parallel"() {
+        given:
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("link.*Executable")
+    }
+
+    def "can execute link shared library tasks in parallel"() {
+        given:
+        withComponentsForAppsAndSharedLibs(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("link.*SharedLibrary")
+    }
+
+    def "can execute create static library tasks in parallel"() {
+        given:
+        withComponentsForAppsAndStaticLibs(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("create.*StaticLibrary")
+    }
+
+    def "can execute compile tasks in parallel"() {
+        given:
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("compile.*")
+    }
+
+    def "can execute assemble tasks in parallel"() {
+        given:
+        apps = [
+            assemblerWithC1 : new AssemblerWithCHelloWorldApp(toolChain),
+            assemblerWithC2 : new AssemblerWithCHelloWorldApp(toolChain)
+        ]
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("assembleAssembler.*")
+    }
+
+    @RequiresInstalledToolChain(VISUALCPP)
+    def "can execute windows resource compile tasks in parallel"() {
+        given:
+        apps = [
+            windowsRC1 : new WindowsResourceHelloWorldApp(),
+            windowsRC2 : new WindowsResourceHelloWorldApp()
+        ]
+        withComponentsForApps(apps)
+
+        when:
+        succeeds("assemble")
+
+        then:
+        assertTasksAreParallel("compile.*Rc")
+    }
+
+    /**
+     * This is an approximation to check that a set of tasks execute in parallel by default.  Basically it checks to
+     * see if any of the tasks execute concurrently with any other task.  We can't guarantee that another task will
+     * always run concurrently while a given task is executing, so we settle for checking that the opposite does not
+     * happen (i.e. that none of the tasks execute concurrently with other tasks).
+     */
+    void assertTasksAreParallel(String regex) {
```

Comment:
I've changed this so that it is similar to the swift test and we put a synthetic task to run parallel to the given task and then validate that there was parallelism.  This should be a little better now.